sidecar/src/figwheel_sidecar/components/figwheel_server.clj

```diff
@@ -152,25 +158,95 @@
                        trace
                        "</pre>")}))))))
 
-(defn handle-index [handler root]
+(defn checksum-file
+  "Generated from Pandect"
+  ;; TODO: full attribution^^
+  [^File file]
+  (with-open [is (io/input-stream file)]
+    (let [buffer-size (int 2048)
+          ba (byte-array buffer-size)
+          adler-32 (new Adler32)]
+      (loop []
+        (let [num-bytes-read (.read is ba 0 buffer-size)]
+          (when-not (= num-bytes-read -1)
+            (.update adler-32 ba 0 num-bytes-read)
+            (recur))))
+      (.getValue adler-32))))
+
+(defn supports-extended-attributes?
+  "Not all filesystems suport Java's UserDefinedFileAttributes (a.k.a. extended attributes),
+  notably HFS+ and APFS on macOS.
+
+  Waiting for https://bugs.openjdk.java.net/browse/JDK-8030048 to add macOS support."
+  [^Path path]
+  (.supportsFileAttributeView
+    (Files/getFileStore path)
+    ^Class UserDefinedFileAttributeView))
+
+(defn ^UserDefinedFileAttributeView get-user-defined-attribute-view [path]
+  (Files/getFileAttributeView
+    path
+    UserDefinedFileAttributeView
+    (into-array LinkOption [])))
+
+(def checksum-attribute-name "user.figwheel.checksum")
+
+(defn get-attribute [path attribute]
+  (try
+    (let [view (get-user-defined-attribute-view path)
+          name attribute
+          size (.size view name)
+          attr-buf (ByteBuffer/allocate size)]
+      (.read view name attr-buf)
+      (.flip attr-buf)
+      (str (.decode (Charset/defaultCharset) attr-buf)))
+    (catch NoSuchFileException e
+      ;; TODO: check up on this
+      nil)))
+
+;; TODO: check if setting an attribute updates file modified time, that would
+;; mess things up.
+(defn set-attribute [path attribute ^String value]
+  (let [view (get-user-defined-attribute-view (.toPath file))]
+    (.write view attribute (.encode (Charset/defaultCharset) value))))
+
+(defn add-file-etag
+  "Adds an etag for a Ring response which contains a File as the body."
+  [response extended-attributes?]
+  (let [file (:body response)]
+    ;; Check if this is a file?
+    (if (instance? File file)
+      (let [path (.toPath file)]
+        (if extended-attributes?
+          (if-let [checksum (get-attribute path checksum-attribute-name)]
+            (response/header response "ETag" checksum)
+            (let [checksum (checksum-file file)]
+              (set-attribute path checksum-attribute-name checksum)
+              (response/header response "ETag" checksum)))
+          (response/header response "ETag" (checksum-file file))))
+      response)))
+
+(defn handle-index [handler root extended-attributes?]
   (fn [request]
     (if (= [:get "/"] ((juxt :request-method :uri) request))
-      (if-let [resp (some-> (resource-response "index.html" {:root (or root "public")
+      (if-let [resp (some-> (resource-response "index.html" {:root root
```

Comment:
I used `root` here because I think the web server root is already calculated if missing at L362?

Reply:
Let's leave it as it isn't part of this commit.